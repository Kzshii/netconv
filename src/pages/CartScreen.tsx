import React, { FC, useState, useEffect } from 'react';
import {
  StyleSheet,
  View,
  SafeAreaView,
  Button,
  Text,
  FlatList,
} from 'react-native';
import { useDispatch } from 'react-redux';
import { useSelector } from 'react-redux';
import { useNavigation } from '@react-navigation/native';
import { ProductsActions } from '../store/actions/products';
import { productsSelector } from '../store/selectors';
import { CartItem } from '../components/CartItem';
import { Product } from '../types';

export const CartScreen: FC = () => {
  const [total, setTotal] = useState(0);
  const dispatch = useDispatch();
  const products = useSelector(productsSelector);
  const navigation = useNavigation();

  useEffect(() => {
    const total = products.reduce((acc, curr) => {
      if (curr.quantity) {
        return acc + curr.price * curr.quantity;
      }
      return acc;
    }, 0);
    setTotal(Math.round(total * 100) / 100);
  }, [products]);

  const onFinish = () => {
    dispatch(ProductsActions.wipeList());
    navigation.navigate('Home');
  };
  const onUpdateProduct = (product: Product) => {
    dispatch(ProductsActions.setProduct(product));
  };

  return (
    <SafeAreaView style={{ flex: 1 }}>
      <View style={styles.container}>
        <FlatList
          data={products.sort((a, b) => a.id - b.id)}
          keyExtractor={(item, idx) => item.name + idx}
          renderItem={({ item }) => (
            <CartItem product={item} setItem={onUpdateProduct} />
          )}
        />
      </View>
      <View style={styles.totalContent}>
        <Text style={styles.totalText}>Total:</Text>
        <Text style={styles.totalText}>R$ {total.toLocaleString('br-BR')}</Text>
      </View>
      <View style={styles.finishButton}>
        <Button color="#fff" title="Finalizar pedido" onPress={onFinish} />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  totalContent: {
    justifyContent: 'space-between',
    flexDirection: 'row',
  },
  totalText: {
    padding: 12,
    fontWeight: 'bold',
    fontSize: 22,
  },
  finishButton: {
    backgroundColor: '#E24E26',
    justifyContent: 'center',
    height: 60,
    bottom: 0,
  },
});
