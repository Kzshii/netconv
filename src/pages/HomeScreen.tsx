import React, { FC } from 'react';
import { StyleSheet, SafeAreaView, FlatList } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { Item } from '../components';
import { Product } from '../types';
import { Data } from '../constants';

export const HomeScreen: FC = () => {
  const navigation = useNavigation();
  const onItemClick = (product: Product) => {
    navigation.navigate('Description', { product });
  };
  return (
    <SafeAreaView style={styles.container}>
      <FlatList
        data={Data}
        numColumns={3}
        keyExtractor={(item, idx) => item.name + idx}
        renderItem={({ item }) => <Item onPress={onItemClick} product={item} />}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
