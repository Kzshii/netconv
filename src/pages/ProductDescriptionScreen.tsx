import React, { FC, useState } from 'react';
import { StyleSheet, View, Text, Button, SafeAreaView } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { useDispatch } from '../hooks/react-hooks';
import { ProductsActions } from '../store/actions/products';
import { Icon } from '../components';
import { Product } from '../types';
import { QuantityItem } from '../components/QuantityItem';

interface Props {
  route: any;
}

export const ProductDescriptionScreen: FC<Props> = ({ route }) => {
  const { product } = route.params;
  const [item, setItem] = useState<Product>(
    new Product({ ...product, quantity: 1 })
  );
  const dispatch = useDispatch();
  const navigation = useNavigation();
  const onAdd = () => {
    dispatch(ProductsActions.setProduct(item));
    navigation.navigate('Home');
  };

  return (
    <SafeAreaView style={{ flex: 1 }}>
      <View style={styles.container}>
        <Icon style={styles.image} path={{ uri: item.image }} />
        <View style={styles.descriptionContent}>
          <Text style={styles.text}>{item.name}</Text>
          <Text style={styles.price}>
            R$ {item.price.toLocaleString('br-BR')}
          </Text>
        </View>
        <QuantityItem product={item} setItem={setItem} />
      </View>
      <View style={styles.addButton}>
        <Button color="#fff" title="Adicionar produto" onPress={onAdd} />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  descriptionContent: {
    marginTop: 32,
    marginBottom: 16,
    paddingLeft: 12,
    paddingRight: 12,
  },
  image: {
    width: '100%',
    height: 200,
    marginBottom: 5,
  },
  text: {
    textAlign: 'center',
    fontSize: 22,
    fontWeight: 'bold',
  },
  price: {
    textAlign: 'center',
    fontSize: 30,
    marginTop: 10,
    fontWeight: 'bold',
  },
  addButton: {
    backgroundColor: '#E24E26',
    justifyContent: 'center',
    height: 60,
    bottom: 0,
  },
});
