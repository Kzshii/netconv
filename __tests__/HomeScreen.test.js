import 'react-native';
import React from 'react';
import { HomeScreen } from '../src/pages/HomeScreen';
import { shallow } from 'enzyme';

jest.mock('@react-navigation/native');

describe('<HomeScreen />', () => {
  test('renders correctly HomeScreen', () => {
    expect(shallow(<HomeScreen />)).toMatchSnapshot();
  });
});
