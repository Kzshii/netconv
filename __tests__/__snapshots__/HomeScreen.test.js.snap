// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen /> renders correctly HomeScreen 1`] = `ShallowWrapper {}`;
